crates/ecs/src/systems/mod.rs

```diff
@@ -408,7 +419,80 @@ impl<'a, Component> DerefMut for Write<'a, Component> {
     }
 }
 
+/// `Query` allows a system to access components from entities other than the currently iterated.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::With;
+/// # use ecs::systems::{Query, Read, Write};
+/// # #[derive(Debug)] struct Mass(f32);
+/// # #[derive(Debug)] struct Velocity(f32);
+/// fn print_largest_momentum(query: Query<(Read<Mass>, Read<Velocity>)>) {
+///     let mut largest_momentum = 0.0;
+///     for (mass, velocity) in query {
+///         let momentum = mass.0 * velocity.0;
+///         if momentum > largest_momentum {
+///             largest_momentum = momentum;
+///         }
+///     }
+///     println!("The largest momentum of all entities is {largest_momentum} kg*m/s.");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Query<'a, P: SystemParameters> {
+    phantom: PhantomData<P>,
+    world: &'a World,
+}
+
+/// Iterator for [`Query`].
+#[derive(Debug)]
+pub struct QueryIterator<'components, P: SystemParameters> {
+    borrowed: P::BorrowedData<'components>,
+    iterate_over_entities: bool,
+    iterated_once: bool,
+}
+
+impl<'a, P: SystemParameters> SystemParameter for Query<'a, P> {
+    type BorrowedData<'components> = &'components World;
+
+    fn borrow<'world>(
+        world: &'world World,
+        _: &[ArchetypeIndex],
+    ) -> SystemParameterResult<Self::BorrowedData<'world>> {
+        Ok(world)
+    }
+
+    unsafe fn fetch_parameter(borrowed: &mut Self::BorrowedData<'_>) -> Option<Option<Self>> {
+        #[allow(trivial_casts)]
+        let world = &*(*borrowed as *const World);
```

Comment:
There is no problem with `world` because the world reference will always be valid then the system is running.
But this `QueryIterator` introduces a new issue with the references in `Read`/`Write` when creating a collection from the iterator.

Look at this example:
```rust
fn system(query: Query<(Read<A>,)>) {
    let vec: Vec<_> = query.into_iter().collect();
    // Here QueryIterator is dropped, so the components_vecs are no longer locked.
    // We cannot prevent the user from dropping the borrowed data when it is inside the iterator.

    for (a,) in vec {
        let x = a.0;
        // We can still access the reference to the components in the component_vec while it is no longer locked.
        // In this case, the scheduler is only thing that prevents potential data races here.
    }
}
```

Reply:
Hm, that's worrying. I would imagine that the `RwLockReadGuard` would prevent us from keeping a reference to its contents before it is dropped, but we somehow bypass that? Is it as a result of our use of `unsafe`? I suppose so, since the lock guard relies on the lifetimes to enforce its invariants, and when we artificially extend the lifetime of `borrowed` it loses track of the data. 

Are you sure there's no clone or copy happening somewhere in the chain that prevents this from being a data race? Or is it an actual reference to the exact same component vec?

> // In this case, the scheduler is only thing that prevents potential data races here.

I mean, maybe that's fine. We did discuss the theoretical redundancy of the locks at some point, since if the scheduler is behaving correctly then they shouldn't be needed. Though, I think actually removing them in practice would involve a lot more `unsafe` code.

Reply:
>  Is it as a result of our use of unsafe?

Yes, this is a result of the `unsafe` usage in Read, where the lifetime of `borrowed` is replaced with the lifetime of `Read`. Before, we could enforce that `borrowed` always outlives `Read`.

> Are you sure there's no clone or copy happening somewhere in the chain that prevents this from being a data race? Or is it an actual reference to the exact same component vec?

The reference in `Read` is pointing to the data in the component vec. Components cannot be copied because we do not require `Clone`.

Reply:
Then I guess there's not much we can do about this, well, other than redesign the entire thing to avoid this. 

For now (aka forever) we'll leave it as-is, since a correctly functioning scheduler should make this a non-issue. But you should add comments in the relevant places mentioning this problem and what can cause it/how to avoid it.